test/util/zip.js

```diff
@@ -23,50 +61,42 @@ const zipStreamToFiles = (zipStream, callback) => {
         yauzl.open(tmpfile, { autoClose: false }, (err, zipfile) => {
-          // eslint-disable-next-line keyword-spacing
-          if(err) return callback(err);
-
-          const result = { filenames: [] };
-          const entries = [];
-          let completed = 0;
-
-          zipfile.on('entry', (entry) => entries.push(entry));
-          zipfile.on('end', () => {
-            if (entries.length === 0) {
-              callback(null, result);
-              zipfile.close();
-            } else {
-              entries.forEach((entry) => {
-                result.filenames.push(entry.fileName);
-                // eslint-disable-next-line no-shadow
-                zipfile.openReadStream(entry, (err, resultStream) => {
-                  // eslint-disable-next-line keyword-spacing
-                  if(err) return callback(err);
-
-                  // eslint-disable-next-line no-shadow
-                  resultStream.pipe(streamTest.toText((err, contents) => {
-                    // eslint-disable-next-line keyword-spacing
-                    if(err) return callback(err);
-
-                    result[entry.fileName] = contents;
-                    completed += 1;
-                    if (completed === entries.length) {
-                      callback(null, result);
-                      zipfile.close();
-                    }
-                  }));
-                });
-              });
-            }
-          });
+          if (err) return callback(err);
+          processZipFile(zipfile, callback);
         });
       }, 5); // otherwise sometimes the file doesn't fully drain
     });
   });
 };
 
-// eslint-disable-next-line no-confusing-arrow
-const pZipStreamToFiles = (zipStream) => new Promise((resolve, reject) => { zipStreamToFiles(zipStream, (err, result) => err ? reject(err) : resolve(result)); });
+// Work around a subtle bug in streams that is not fully understood.  Probably one of:
+//
+// * the cleanup code in test/util/zip.js, or
+// * an unfulfilled interaction that supertest is expecting when working with streams, or
+// * a bug in supertest's pipe/stream implementation
+//
+// Possibly: https://github.com/ladjs/supertest/issues/487
+const binaryParser = (res, callback) => {
+  res.setEncoding('binary');
+  let data = '';
+  res.on('data', (chunk) => { data += chunk; });
+  res.on('end', () => {
+    callback(null, Buffer.from(data, 'binary'));
+  });
+};
+
+const httpZipResponseToFiles = (zipHttpResponse) => new Promise((resolve, reject) => {
+  zipHttpResponse.buffer().parse(binaryParser).end((err, res) => {
+    if (err) return reject(err);
 
-module.exports = { zipStreamToFiles, pZipStreamToFiles };
+    // eslint-disable-next-line no-shadow
+    yauzl.fromBuffer(res.body, (err, zipfile) => {
+      if (err) return reject(err);
+
+      // eslint-disable-next-line no-shadow
+      processZipFile(zipfile, (err, result) => { if (err) reject(err); else resolve(result); });
```

Comment:
Would you mind making the `if; else;` two lines or preferably four? I'm not ready yet for both on one line. 😅